src/back/spv/writer.rs

```diff
@@ -1318,6 +1400,494 @@ impl Writer {
+                    let argument = &ir_function.arguments[index as usize];
+                    let last_idx = last_member_idx(argument.ty, ir_module)?;
+                    let parameter_id = function.parameter_id(index);
+                    (parameter_id, last_idx)
+                }
+                _ => return Err(Error::Validation("array length expression")),
+            },
+            _ => return Err(Error::Validation("array length expression")),
+        };
+
+        let length_id = self.id_gen.next();
+        block.body.push(Instruction::array_length(
+            self.get_uint_type_id()?,
+            length_id,
+            structure_id,
+            member_idx,
+        ));
+
+        Ok(length_id)
+    }
+
+    /// Compute the length of a subscriptable value.
+    ///
+    /// Given `sequence`, an expression referring to some indexable type, return
+    /// its length. The result may either be computed by SPIR-V instructions, or
+    /// known at shader translation time.
+    ///
+    /// `sequence` may be a `Vector`, `Matrix`, or `Array`, a `Pointer` to any
+    /// of those, or a `ValuePointer`. An array may be fixed-size, dynamically
+    /// sized, or use a specializable constant as its length.
+    fn write_sequence_length(
+        &mut self,
+        sequence: Handle<crate::Expression>,
+        ir_module: &crate::Module,
+        ir_function: &crate::Function,
+        fun_info: &FunctionInfo,
+        function: &Function,
+        block: &mut Block,
+    ) -> Result<KnownOrComputed<u32>, Error> {
+        let sequence_ty = fun_info[sequence].ty.inner_with(&ir_module.types);
+        match sequence_ty.indexable_length(ir_module)? {
+            crate::proc::IndexableLength::Known(known_length) => {
+                Ok(KnownOrComputed::Known(known_length))
+            }
+            crate::proc::IndexableLength::Dynamic => {
+                let length_id = self.write_runtime_array_length(
+                    sequence,
+                    ir_module,
+                    ir_function,
+                    function,
+                    block,
+                )?;
+                Ok(KnownOrComputed::Computed(length_id))
+            }
+            crate::proc::IndexableLength::Specializable(constant) => {
+                let length_id = self.constant_ids[constant.index()];
+                Ok(KnownOrComputed::Computed(length_id))
+            }
+        }
+    }
+
+    /// Compute the maximum valid index of a subscriptable value.
+    ///
+    /// Given `sequence`, an expression referring to some indexable type, return
+    /// its maximum valid index - one less than its length. The result may
+    /// either be computed, or known at shader translation time.
+    ///
+    /// `sequence` may be a `Vector`, `Matrix`, or `Array`, a `Pointer` to any
+    /// of those, or a `ValuePointer`. An array may be fixed-size, dynamically
+    /// sized, or use a specializable constant as its length.
+    fn write_sequence_max_index(
```

Comment:
To avoid leaking a bunch of `writer`-local types, this needed to be a sub-module of `writer`: `pub(in ...)` can only restrict visibility to parents of the module containing the definition. So it's `src/back/spv/writer/index.rs`, which is a bit deep, but not the end of the world.

Reply:
If making `index` a child module of `writer` is too much nesting, then there's the alternative of adding a bunch of non-`pub` `use` declarations to `src/back/spv/mod.rs`. I didn't explore that possibility.

Reply:
yeah, I don't think we need this level of nesting in either parts of Naga.
Writer isn't meant to be this huge sub-module of SPV-out. It's just accidental that it has all the logic today. Coincidentally, "writer-local" is pretty much "spv-in" namespace, since "writer" dominates it. So `pub(super)` would work ok here.